### DISCUSSION TITLE
unnecessary injection of css when   "clientRouting: false"

### DISCUSSION THREAD

**Discussion by DmytroZozuliak** [vikejs/vike#1785] [Discussion]
I'm using "clientRouting: false" parameter in my `+config.js` file.
So i no need in such files (hooks) as `"+onPageTransitionStart.js"` and `"+onPageTransitionEnd.js".`
but i figure out that css is still injecting in  "/dist/client/assets/static/" folder (if hooks are present, css file is `"onPageTransitionStart.Db35thue.css"`, but when i remove that hooks - css file becomes `"LayoutDefault.CNGbaUCn.css"`). So i get two almost identical css files `"LayoutDefault.someHash.css"` which injects in head tag of body when i'm deploying pages.
And when i'm using tailwind all of that classes are duplicated.
![image](https://github.com/user-attachments/assets/f78d2018-cd8e-4d36-ae24-50fbc6f15ce5)

**Comment by DmytroZozuliak** [comment]
Update:

I used the `disableAutoFullBuild `option in` vite.config.ts`:
```
vike({
    disableAutoFullBuild: true
  }),
```
I then built the client and server separately using the commands in the documentation.

When building the client, everything works as expected (only one CSS file appears in the `dist/client/assets/static` folder). However, when building the server, another CSS file appears in the `dist/server/assets/static` folder, and also in the `dist/client/assets/static` folder (both with the same hash).

Is this the normal flow? How can I avoid this duplication?

Additionally, if I haven't previously built the client, the server build produces an error.

![image](https://github.com/user-attachments/assets/57ba3d2b-0b77-4a08-8416-70e7e4601495)

**Reply by brillout** [reply]
The `disableAutoFullBuild` option will likely be removed once Vike's CLI replaces Vite's CLI (ETA this month). So, ideally, we should make it work without having to use this option.

Can you create a minimal reproduction of your original issue? With steps to reproduce and "what I expect" + "what I get instead" descriptions.

**Comment by DmytroZozuliak** [comment]
created new app using   [Bati](https://batijs.dev/)   with 

`pnpm create bati --react --tailwindcss --fastify --eslint --prettier`

Changed :
1.  added   `clientRouting: false` in `+config.ts`,
2. removed `usePageContext();` in Link component,
3. deleted `+onPageTransitionStart.ts`, `+onPageTransitionEnd.ts` files

All of that changes related to non clientRouting in my app

After that built project with build command (`vite build`)

We are only interesting of the result of building css files.
<img width="385" alt="Screenshot 2024-08-05 at 15 00 30" src="https://github.com/user-attachments/assets/b370f6a0-4568-4075-b3b8-95f9b0ba7dd9">
One of css files in `dist/client/assets/static` folder  creates when building client, and another css file in that folder creates when building server.

It's easily can be reproduce by setting `disableAutoFullBuild: true,` in `vite.config.ts` file and firing one by one `vite build` and `vite build --ssr`.

Because of creation that separate almost similar css files in `dist/client/assets/static` caused to inject them both in `<head>` tag

![image](https://github.com/user-attachments/assets/1b5f2c9e-8bbc-415d-a7f9-9120f2f2f514)

And if my css files are huge it will cause a big problem of first time render page value.

I dont't know why after building server part also creates  css file in `dist/client/assets/static` (my purpose is to add some classes whith changing pages with clientRouting ), but in such case i'm not using clientRouting at all.

So i expect that after building my project there will not be additional injection of css.

project with changes can be reached in
https://stackblitz.com/edit/in627q?file=README.md,components%2FLink.tsx,pages%2F%2Bconfig.ts

**Reply by brillout** [reply]
You aren't supposed to use `vike-react` with `clientRouting: false`. Why do you want to use `clientRouting: false`?

As for your CSS issues, the best would be a reproduction with vanilla Vike using https://vike.dev/new#without-vike-react-vue-solid.

**Reply by DmytroZozuliak** [reply]
With `clientRouting: false`, my bundle size decreased when downloading only one page, which is crucial for my page speed.

 I replicated the main structure of my app using vanilla Vike, and while the CSS issue is resolved, I’m now dealing with a lot of raw logic that `vike-react` previously handled.

Is there a way to address this in `vike-react`? Alternatively, could you point me to the relevant part of the code so I can work on a PR to resolve this?

To grab your attention, I’ve sent you $50 (for now).

**Reply by brillout** [reply]
> With `clientRouting: false`, my bundle size decreased when downloading only one page, which is crucial for my page speed.

Indeed with `clientRouting: false`, and even more so with HTML-only, you can dramatically decrease client-side KBs.

> I replicated the main structure of my app using vanilla Vike, and while the CSS issue is resolved, I’m now dealing with a lot of raw logic that `vike-react` previously handled.
> 
> Is there a way to address this in `vike-react`? Alternatively, could you point me to the relevant part of the code so I can work on a PR to resolve this?

I'm a bit reluctant to make `clientRouting: false` work with `vike-react` because it complexifies `vike-react`. Ejecting `vike-react` like you're doing is actually very much something we want to encourage. The source code of `vike-react` is small and relatively simple, so I'm inclined to think that you'll be fine cherry-picking & replicating some of vike-react's features.

> To grab your attention, I’ve sent you $50 (for now).

Thank you :green_heart:

If your company is up for long-term sponsorship, then we're open for opportunities to collaborate. We also offer collaboration on a one-time contractual basis.
